enigma/src/main/java/cuchaz/enigma/classhandle/ClassHandleProvider.java

```diff
@@ -300,7 +300,11 @@ private void continueMapSource(CompletableFuture<Result<DecompiledClassSource, C
 				Entry.this.waitingSources.forEach(s -> s.complete(source));
 				Entry.this.waitingSources.clear();
 				withLock(lock.readLock(), () -> new ArrayList<>(handles)).forEach(h -> h.onMappedSourceChanged(source));
-			}, p.pool);
+			}, p.pool)
+			.exceptionally(exc -> {
+				exc.printStackTrace();
```

Comment:
Any exceptions should be caught inside the thenApply lambdas so they get displayed in the GUI. For that, catch the exception where it occurs (in continueIndexSource, I assume) and turn into a ClassHandleError so that it can get propagated correctly. Might be worth adding a new ClassHandleError.Type for indexing, I wasn't aware that errors could occur during that.

Reply:
This isn't really how CompletableFuture is supposed to be used, it itself is already a result+exc carrier, handles exception fall-through and captures every exception reliably. I think the proper solution there is to remove `Result` and replace the 2nd part of `thenAcceptAsync` with `whenCompleteAsync`.

This is outside the scope of the PR though, so just printing the exception seems fair for unexpected issues like this until this has been refactored.

Trying to catch the exceptions everywhere manually is ineffective and doomed to miss something eventually.

Reply:
If you look at how it's used in continueMapSource it isn't easily changable without rewriting how this entire thing works: the Result gets saved and passed around elsewhere. Making a second code path just sounds like unnecessary effort. I see exceptionally() lets you return something though so returning the Result.err there should probably work, but then it should be done in `continueIndexSource`. Returning null as you do here means that the currently executing future is "out of date" because the versions mismatch (e.g. user changed a mapping while it was decompiling) and isn't as this is supposed to work. It will just create futures that will never be completed.

Reply:
We can probably call [`exceptionallyAsync`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/concurrent/CompletionStage.html#exceptionallyAsync(java.util.function.Function,java.util.concurrent.Executor)) since we are on jdk 16.

Reply:
Eh, it's probably not worth submitting a new async task here for something that takes almost no time, especially for something that probably only gets called when an exception happens which should be fairly rare. The Executor overhead there would probably be more than the executed code. (Regardless, after some discussion in PMs this part is already implemented slightly differently in bc1f985)